### ISSUE TITLE
Feature request: Add a wait option to docker stack rm

### ISSUE THREAD

**Issue by kinghuang** [moby/moby#30942]
**Description**

I have several GitLab repositories that deploy review apps to Docker in swarm mode. When done manually, this typically involves the following `docker` commands.

```
docker stack rm my_project
docker stack deploy --compose-file=my_project.yml my_project
```

`docker stack rm` issues the remove command for services and networks and returns right away. When the above two commands are done in rapid succession by a CI job, I often get errors about missing networks. It seems that re-deploying a stack that has just been removed but whose services/networks haven't finished removing can result in errors.

To avoid errors, I've added additional steps between `rm` and `deploy` in my CI scripts to specifically wait for services and networks to finish removing before deploying again.

```
until [ -z "$(docker service ls --filter label=com.docker.stack.namespace=$COMPOSE_PROJECT_NAME -q)" ] || [ "$limit" -lt 0 ]; do
  sleep 1;
done

until [ -z "$(docker network ls --filter label=com.docker.stack.namespace=$COMPOSE_PROJECT_NAME -q)" ] || [ "$limit" -lt 0 ]; do
  sleep 1;
done
```

It would be useful if the `docker stack rm` command had an option like `--wait` that waits for services and networks to be fully removed for a stack before exiting. A `--wait-timeout` option could also be added to specify how long it should wait before exiting if services and networks take a long time to remove.

This would be a bit like the `docker container wait` command, but as an option for removing stacks.

**Steps to reproduce the issue:**

1. Have a stack definition (bundle or compose file) and deploy it.
2. Remove the stack.
3. Immediately deploy the stack again with the same name.

**Describe the results you received:**

`docker` reports errors with missing networks, sometimes. The networks from the previous deployment are actually still in the process of being removed.

**Describe the results you expected:**

The stack should be redeployed.

**Additional information you deem important (e.g. issue happens only occasionally):**

This is most prevalent with large stacks. Networks are removed after all the services in them have stopped. This can take a little while.

**Output of `docker version`:**

```
(swarm) →  ~ docker version
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 08:47:51 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:38:28 2017
 OS/Arch:      linux/amd64
 Experimental: true

```

**Output of `docker info`:**

```
(swarm) →  ~ docker info
Containers: 14
 Running: 14
 Paused: 0
 Stopped: 0
Images: 71
Server Version: 1.13.1
Storage Driver: devicemapper
 Pool Name: docker-thinpool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 23.1 GB
 Data Space Total: 65.28 GB
 Data Space Available: 42.18 GB
 Metadata Space Used: 6.316 MB
 Metadata Space Total: 683.7 MB
 Metadata Space Available: 677.4 MB
 Thin Pool Minimum Free Space: 6.527 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: active
 NodeID: 3vz5vqb96ab3yh6mrhhbnf5ki
 Is Manager: true
 ClusterID: 9p9qjvvrp991zitpixlm8dpm5
 Managers: 4
 Nodes: 4
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 10.45.32.40
 Manager Addresses:
  10.45.32.40:2377
  10.45.32.41:2377
  10.45.32.42:2377
  10.45.32.43:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.6.1.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.3 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.51 GiB
Name: itpocnode01.ucalgary.ca
ID: OM7K:NNN7:75VJ:I34C:W5ZS:TCYH:ZGPH:SRXP:HINS:7O2W:IZQ4:QIGD
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: kinghuang
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by jaguardev** [comment]
+1 requesting this feature
faced exactly same thing

@kinghuang thanks for workaround!

**Comment by samattridge** [comment]
I've just come across this issue too so thanks to @kinghuang for the workaround.  I would like to see this option on the 'docker stack create' command as well.

**Comment by Hermain** [comment]
+1

**Comment by dustin-decker** [comment]
As mentioned in https://github.com/moby/moby/issues/32367#issuecomment-301908838 adding a `--detach=false` flag would provide this functionality and interactively show progress on what parts of the stack have been removed.

**Comment by vasicvuk** [comment]
+1

**Comment by majstorki88** [comment]
+1

**Comment by sophieschwalbe** [comment]
+1

**Comment by brandshaide** [comment]
+1 And thanks, @kinghuang for sharing your nice workaround 👍

**Comment by joeynovak** [comment]
+1 would love to have this

**Comment by kinghuang** [comment]
@Dalzhim Yeah, there's actually an extra sleep after the two loops in some of my CI jobs. You should be careful about using `docker ps` instead of `docker service`, though. That will only include containers on the specific node that you're communicating with. That'll only be valid if your swarm consists of a single node.

An alternative to that might be to loop over all the nodes in the swarm and check the tasks on each node using `docker node ps …`. I've found it sufficient to just wait a few extra seconds in my clusters.

**Comment by Dalzhim** [comment]
@kinghuang I had overlooked the multi node scenario, thanks for pointing that out! Now `docker service ls` does sound better when the objective is to redeploy. In the case of continuous integration, I want to prevent conflicts with `docker build` so I think using `docker ps` is appropriate for this particular use case. I guess I'm going to use 3 loops then!

**Comment by thaJeztah** [comment]
I think this will have to be implemented in the CLI, so let's close this in favour of https://github.com/docker/cli/issues/373 (please continue the discussion over there 👍)

**Comment by positron96** [comment]
Hello everyone! 
While we wait on a correct solution, I have a question about a proposed workaround - what's with `[ "$limit" -lt 0 ]` part? Do I have to define some limit variable and decrement it manually?

**Comment by halis** [comment]
> Hello everyone!
> While we wait on a correct solution, I have a question about a proposed workaround - what's with `[ "$limit" -lt 0 ]` part? Do I have to define some limit variable and decrement it manually?

Yeah wondering the same thing myself.

**Comment by halis** [comment]
Gotcha, thanks for the information!

I think we are running into this same issue and once you deploy too soon, with a dangling network out there, pretty much everything is hosed until you restart the docker daemon.

I feel like Swarm should implement a queue where they track stack removals and prevent stack deploys until the stack removals are finished.

🤷‍♂️

**Comment by miltoncs** [comment]
I prefer the command
`until [ -z $(docker stack ps {{ stack_name }} -q) ]; do sleep 1; done`
We use it in an ansible playbook!

**Comment by thaJeztah** [comment]
I see the discussion on this ticket is still ongoing, but please continue in the cli repository instead; https://github.com/moby/moby/issues/30942#issuecomment-387061567

> I think this will have to be implemented in the CLI, so let's close this in favour of https://github.com/docker/cli/issues/373 (please continue the discussion over there 👍)

locking this issue so that the discussion is done in a single place
